### ISSUE TITLE
[FR/Discussion] new name primitive with effectful string formatting

### ISSUE THREAD

**Issue by ordabayevy** [pyro-ppl/pyro#2713]
This relates to the manipulation of the `msg["name"]` by `VectorizedMarkovMessenger`. In order to build the `step` information (e.g., `("x_0", "x_prev", "x_curr")` `VectorizedMarkovMessenger` manipulates the suffix of the `msg["name"]` assuming that variables have following names: `pyro.sample("x_{}".format(i) ...)`. This implementation is limiting and puts constraints on how the variables must be named.

Making `str.format` method effectful would give greater flexibility in string manipulation by processing its arguments before the `.format` method is called. Since Python built-in types cannot be altered, a new `name` primitive could be added that subclasses built-in `str` type:

```py
class name(str):
    pass

name.format = effectful(name.format, type="format")

...

def model():
    pyro.sample(name("x_{}").format(i), ...)

...


def _pyro_format(self, msg):
    # process msg["args"]
```

Another use case for `name` primitive can be in name scoping for `pyro.sample` and `pyro.param`. @eb8680 do you think this might work?

**Comment by eb8680** [comment]
@ordabayevy interesting idea! To clarify, you're suggesting using this in `VectorizedMarkovMessenger` to build `step` by using the unformatted name (`"x_{}"`) to align sample sites across timesteps?

Another possible interface in place of your `name` class would involve delegating name formatting to an autonaming handler, so that the `.format` call never appears in user code in the first place.  A very crude version of this is implemented in [`pyro.contrib.autoname.scope`](http://docs.pyro.ai/en/dev/contrib.autoname.html), but it doesn't know anything about `pyro.markov`.

A more general autonaming handler would implement the algorithm for name generation in Section 2.2 of the paper ["Lightweight implementations of probabilistic programming languages via transformational compilation"](http://www.jmlr.org/proceedings/papers/v15/wingate11a/wingate11a.pdf), which operates on a similar principle to the `DimStack` in `pyro.contrib.funsor` - it keeps track of a stack of all of the active function calls, loops and iterator states (or Markov contexts, in Pyro) and uses that information to format names intelligently.

Here's an example of what a minimal syntax might look like:
```py
@markov_autoname
@pyro.markov
def model(...):
    ...
    for i in vectorized_markov(range(10), name="time"):
        x = pyro.sample("x", Bernoulli(p[x]))  # automatically format "x" to e.g. "model__x__time_{i}"
    ...
```
We could still use your idea for inferring `step` from the unformatted name `"x"` here inside `vectorized_markov`.

**Comment by ordabayevy** [comment]
> To clarify, you're suggesting using this in VectorizedMarkovMessenger to build step by using the unformatted name ("x_{}") to align sample sites across timesteps?

Yeah, though I haven't thought out the details.

> A more general autonaming handler would implement the algorithm for name generation in Section 2.2 of the paper "Lightweight implementations of probabilistic programming languages via transformational compilation", which operates on a similar principle to the DimStack in pyro.contrib.funsor - it keeps track of a stack of all of the active function calls, loops and iterator states (or Markov contexts, in Pyro) and uses that information to format names intelligently

I like this idea a lot. Has anyone worked on it already? Do you mind if take a crack at implementing it?

**Comment by eb8680** [comment]
> I like this idea a lot. Has anyone worked on it already? Do you mind if I take a crack at implementing it?

Sure, nobody else is working on this right now and it's a feature I'd very much like to see in Pyro. We should probably discuss the design a bit first, though - let's cover this on Friday, unless this is blocking some of your other work.

**Comment by ordabayevy** [comment]
Awesome! There is no rush, let's discuss this more on Friday.

**Comment by ordabayevy** [comment]
Addressed by #2745
